### ISSUE TITLE
targetcreated event not firing when 1st parameter of window.open() is empty string

### ISSUE THREAD

**Issue by ThomasSimeroth** [puppeteer/puppeteer#2810]
The targetcreated event is not firing when window.open() is called.

### Steps to reproduce

**Tell us about your environment:**

* Puppeteer version: 1.5.0
* Platform / OS version: Windows 10.0.17134
* Node.js version: 8.11.2

**What steps will reproduce the problem?**

1. `const newPagePromise = new Promise(x => browser.on('targetCreated', target => x(target.page())));`
2.  `await page.click("#bReport"); //Makes popup`
3.  `const popup = await newPagePromise; //Hangs here`

In the website, the button is rigged to call
`var report=window.open("","Schedule","menubar=0,scrollbars=1,resizable=1");`
before adding content via an AJAX call.

**What is the expected result?**
The popup should appear in a new window, with the variable pointing at the new page created by window.open()

**What happens instead?**
The popup does appear, but the promise hangs. I have no way to validate the contents of the popup.

This may be a duplicate of #1992. I'm making a new issue to increase visibility and hopefully find a solution.

**Comment by aslushnikov** [comment]
The [`'targetcreated'`](https://pptr.dev/#?product=Puppeteer&version=v1.5.0&show=api-event-targetcreated) event is spelled with all-small letters; you have a capital 'C'.

A rule of a thumb is that all pptr literals are small-case.

**Comment by ThomasSimeroth** [comment]
@aslushnikov I changed to a lower case c, but the same issue is still occurring. The typo is fairly embarrassing, but it doesn't appear to be the root cause of the problem.

**Comment by aslushnikov** [comment]
@ThomasSimeroth I wasn't able to reproduce; if you have a puppeteer script that I can run locally that would reproduce the problem, I'll be a great help!

**Comment by ThomasSimeroth** [comment]
Dug a little deeper on this one.
This code works:
```js
const newPagePromise = new Promise(x => browser.on('targetcreated', target => x(target.page())));
await page.evaluate(async () => window.open("https://google.com","Schedule","menubar=0,scrollbars=1,resizable=1"));
const popup = await newPagePromise;
const image = await popup.screenshot();

```

However, this fails.
```js
const newPagePromise = new Promise(x => browser.on('targetcreated', target => x(target.page())));
  await page.evaluate(async () => await window.open("","Schedule","menubar=0,scrollbars=1,resizable=1"));
const popup = await newPagePromise;
const image = await popup.screenshot();
```

The issue appears to be leaving the first field of window.open blank. Running the script in headful mode does create a blank popup, but the promise fails to resolve. My guess is puppeteer is expecting something in the address field, and is failing when it gets an empty string. For instance, typing in a random string in the first parameter, such as "afd", will allow the promise to resolve, with a resource not found notification appearing in the popup.

**Comment by aslushnikov** [comment]
@ThomasSimeroth thanks! Indeed, I was able to reproduce this with the following script:

```js
const pptr = require('puppeteer');

(async () => {
  const browser = await pptr.launch({headless: false});
  const page = await browser.newPage();
  const newPagePromise = new Promise(x => browser.on('targetcreated', target => x(target.page())));
  await page.evaluate(async () => await 
  window.open("","Schedule","menubar=0,scrollbars=1,resizable=1"));
  const popup = await newPagePromise;
  console.log('Done!');
  await browser.close();
})();
```

@JoelEinbinder: the `target._isInitialized` flag is failing here. I wonder why we don't get `targetInfoChanged` event from protocol since the popup is actually pointing to `about:blank`.

**Comment by ThomasSimeroth** [comment]
@JoelEinbinder Any update on this? I realize this might not be high priority, but it's important to me. I'm using Puppeteer for a summer internship, so I would like this fixed by the end of July if possible. I tried fixing it myself, but I didn't make much progress. I'm hoping you can help.

**Comment by ziadali1** [comment]
@ThomasSimeroth I'm having the same issue. Did you find an workaround?

**Comment by ThomasSimeroth** [comment]
@ziadali1 Nope. I had originally replied that changing `window.open("")` to `window.open("about:blank")` in the source code would bypass that issue. While that works fine in Puppeteer, it causes issues with other browsers. More specifically, changing the method in Firefox locks the popup, preventing any content getting written to it. That's obviously an undesirable side effect, so I would not recommend using that workaround.

**Comment by ThomasSimeroth** [comment]
Digging a little deeper, this appears to be a bug in DevTools Protocol, rather than Puppeteer.  Via some console.logs(), I've discovered that the `targetInfoChanged` event is firing for the new window. This is where things get strange though. The `targetInfo` being stored before the event is run has the url being "", as expected. However, the `targetInfo` being passed in as a parameter, also has `url: ""`. 

In fact, the `targetInfo` being passed in is identical to the one in storage. This begs the question of why `targetInfoChanged` is firing when the targetInfo hasn't changed! Clearly, something is amiss here. Whether the Protocol is sending an old version of the object or if it's failing to write the object properly is unclear, but it helps narrow down the root cause.

**Comment by ziadali1** [comment]
@ThomasSimeroth Unfortunately i do not have access to the source code :(

**Comment by ziadali1** [comment]
I noticied that when a new window is created with some link, like this: 

`await page.evaluate(async () => await window.open("http://google.com","Schedule","menubar=0,scrollbars=1,resizable=1"));`

the code follows to these functions:

_targetCreated
_targetInfoChanged
_targetInfoChanged

However, when a new window is created with auto:blank

`await page.evaluate(async () => await window.open("","Schedule","menubar=0,scrollbars=1,resizable=1"));`

this is the flow:

_targetCreated
_targetInfoChanged


So when i click a link that i know is auto:blank i search for a target that is not initialized `'target._isInitialized`' and call the `_targetInfoChanged` again passing the targetInfo as parameter.
It's working for now. This is just an workaround until the bug is fixed.

**Comment by airen29** [comment]
This issue can be also replicated easily if you have website that is using Facebook login. You request Facebook Login => website opens new separate window => window is not registered (and it doesn't open a window with ""). 

Any info how to deal with that issue? Thanks

**Comment by shobirin-web** [comment]
> I noticied that when a new window is created with some link, like this:
> 
> `await page.evaluate(async () => await window.open("http://google.com","Schedule","menubar=0,scrollbars=1,resizable=1"));`
> 
> the code follows to these functions:
> 
> _targetCreated
> _targetInfoChanged
> _targetInfoChanged
> 
> However, when a new window is created with auto:blank
> 
> `await page.evaluate(async () => await window.open("","Schedule","menubar=0,scrollbars=1,resizable=1"));`
> 
> this is the flow:
> 
> _targetCreated
> _targetInfoChanged
> 
> So when i click a link that i know is auto:blank i search for a target that is not initialized `'target._isInitialized`' and call the `_targetInfoChanged` again passing the targetInfo as parameter.
> It's working for now. This is just an workaround until the bug is fixed.

any sample code for this workaround ?

**Comment by nanthanwa** [comment]
I'm facing the same problem as you guys with the URL `about:blank`.

2 years passed for this issue, has anyone can deal with it?

**Comment by epicfoox** [comment]
Seems like a pretty common issue, I'm also looking for a solution here. Any progress with it?

**Comment by epicfoox** [comment]
.. so I've managed to force `targetcreated` event in `browser` after opening `about:blank` new window by running this function:

```javascript
function forceTargetInitialization(browser) {
    Array.from(browser._targets.values()).forEach((t, i) => {
        if (!t._isInitialized) {
            console.log('Forcing target initialization');
            browser._targetInfoChanged({
                targetInfo: { ...t._targetInfo, url: ' ' },
            });
        }
    });
}
```

I needed to force some random url there just to be different than empty string `''`.

Careful though, if you have uninitialised targets for a reason you might run into troubles there. And make sure you don't need that url that is now obviously wrong.

I would give myself a thumbs down for this solution but it helped and maybe will help someone else too. At least until it's fixed in puppeteer/dtp.
